// @ton/test-utils/dist/test/comparisons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compareSliceForTest = exports.compareAddressForTest = exports.compareCellForTest = void 0;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function compareCellForTest(subject, cmp) {
    return {
        pass: cmp.equals(subject),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        posMessage: ((subject, cmp) => `Expected\n${subject}\nto equal\n${cmp}`).bind(undefined, subject, cmp),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        negMessage: ((subject, cmp) => `Expected\n${subject}\nNOT to equal\n${cmp}\nbut it does`).bind(undefined, subject, cmp),
    };
}
exports.compareCellForTest = compareCellForTest;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function compareAddressForTest(subject, cmp) {
    return {
        pass: cmp.equals(subject),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        posMessage: ((subject, cmp) => `Expected ${subject} to equal ${cmp}`).bind(undefined, subject, cmp),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        negMessage: ((subject, cmp) => `Expected ${subject} NOT to equal ${cmp}, but it does`).bind(undefined, subject, cmp),
    };
}
exports.compareAddressForTest = compareAddressForTest;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function compareSliceForTest(subject, cmp) {
    return {
        pass: cmp.asCell().equals(subject.asCell()),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        posMessage: ((subject, cmp) => `Expected\n${subject}\nto equal\n${cmp}`).bind(undefined, subject, cmp),
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        negMessage: ((subject, cmp) => `Expected\n${subject}\nNOT to equal\n${cmp}\nbut it does`).bind(undefined, subject, cmp),
    };
}
exports.compareSliceForTest = compareSliceForTest;
