// @ton/blueprint/dist/network/send/MnemonicProvider.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _MnemonicProvider_wallet, _MnemonicProvider_secretKey, _MnemonicProvider_client, _MnemonicProvider_ui, _MnemonicProvider_network;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MnemonicProvider = void 0;
const core_1 = require("@ton/core");
const crypto_1 = require("@ton/crypto");
const wallets_1 = require("./wallets");
class MnemonicProvider {
    constructor(params) {
        _MnemonicProvider_wallet.set(this, void 0);
        _MnemonicProvider_secretKey.set(this, void 0);
        _MnemonicProvider_client.set(this, void 0);
        _MnemonicProvider_ui.set(this, void 0);
        _MnemonicProvider_network.set(this, void 0);
        if (!(params.version in wallets_1.wallets)) {
            throw new Error(`Unknown wallet version ${params.version}, expected one of ${Object.keys(wallets_1.wallets).join(', ')}`);
        }
        __classPrivateFieldSet(this, _MnemonicProvider_client, params.client, "f");
        __classPrivateFieldSet(this, _MnemonicProvider_network, params.network, "f");
        const kp = (0, crypto_1.keyPairFromSecretKey)(params.secretKey);
        __classPrivateFieldSet(this, _MnemonicProvider_wallet, (0, core_1.openContract)(this.createWallet(params, kp), (params) => __classPrivateFieldGet(this, _MnemonicProvider_client, "f").provider(params.address, params.init && {
            ...params.init,
            data: params.init.data ?? undefined,
            code: params.init.code ?? undefined,
        })), "f");
        __classPrivateFieldSet(this, _MnemonicProvider_secretKey, kp.secretKey, "f");
        __classPrivateFieldSet(this, _MnemonicProvider_ui, params.ui, "f");
    }
    createWallet(params, kp) {
        if (params.version === 'v5r1') {
            return wallets_1.wallets[params.version].create({
                publicKey: kp.publicKey,
                walletId: {
                    networkGlobalId: params.network === 'testnet' ? -3 : -239, // networkGlobalId: -3 for Testnet, -239 for Mainnet
                    context: {
                        workchain: params.workchain ?? 0,
                        subwalletNumber: params.subwalletNumber ?? 0,
                        walletVersion: 'v5r1',
                    },
                },
            });
        }
        return wallets_1.wallets[params.version].create({
            workchain: params.workchain ?? 0,
            publicKey: kp.publicKey,
            walletId: params.walletId,
        });
    }
    async connect() {
        const formattedAddress = this.address().toString({
            testOnly: __classPrivateFieldGet(this, _MnemonicProvider_network, "f") === 'testnet',
            bounceable: false,
        });
        __classPrivateFieldGet(this, _MnemonicProvider_ui, "f").write(`Connected to wallet at address: ${formattedAddress}\n`);
    }
    async sendTransaction(address, amount, payload, stateInit) {
        await __classPrivateFieldGet(this, _MnemonicProvider_wallet, "f").sendTransfer({
            seqno: await __classPrivateFieldGet(this, _MnemonicProvider_wallet, "f").getSeqno(),
            secretKey: __classPrivateFieldGet(this, _MnemonicProvider_secretKey, "f"),
            messages: [
                {
                    init: stateInit,
                    body: payload ?? new core_1.Cell(),
                    info: {
                        type: 'internal',
                        ihrDisabled: true,
                        ihrFee: 0n,
                        bounce: true,
                        bounced: false,
                        dest: address,
                        value: { coins: amount },
                        forwardFee: 0n,
                        createdAt: 0,
                        createdLt: 0n,
                    },
                },
            ],
            sendMode: core_1.SendMode.PAY_GAS_SEPARATELY,
        });
        __classPrivateFieldGet(this, _MnemonicProvider_ui, "f").write('Sent transaction');
    }
    address() {
        return __classPrivateFieldGet(this, _MnemonicProvider_wallet, "f").address;
    }
}
exports.MnemonicProvider = MnemonicProvider;
_MnemonicProvider_wallet = new WeakMap(), _MnemonicProvider_secretKey = new WeakMap(), _MnemonicProvider_client = new WeakMap(), _MnemonicProvider_ui = new WeakMap(), _MnemonicProvider_network = new WeakMap();
