// @tact-lang/opcode/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dictionary = exports.Cell = exports.debugSymbols = exports.AssemblyWriter = exports.disassembleAndProcess = exports.disassembleRawRoot = exports.disassembleRoot = exports.disassemble = void 0;
var disasm_1 = require("./decompiler/disasm");
Object.defineProperty(exports, "disassemble", { enumerable: true, get: function () { return disasm_1.disassemble; } });
Object.defineProperty(exports, "disassembleRoot", { enumerable: true, get: function () { return disasm_1.disassembleRoot; } });
Object.defineProperty(exports, "disassembleRawRoot", { enumerable: true, get: function () { return disasm_1.disassembleRawRoot; } });
Object.defineProperty(exports, "disassembleAndProcess", { enumerable: true, get: function () { return disasm_1.disassembleAndProcess; } });
var assembly_writer_1 = require("./printer/assembly-writer");
Object.defineProperty(exports, "AssemblyWriter", { enumerable: true, get: function () { return assembly_writer_1.AssemblyWriter; } });
var known_methods_1 = require("./utils/known-methods");
Object.defineProperty(exports, "debugSymbols", { enumerable: true, get: function () { return known_methods_1.debugSymbols; } });
var core_1 = require("@ton/core");
Object.defineProperty(exports, "Cell", { enumerable: true, get: function () { return core_1.Cell; } });
Object.defineProperty(exports, "Dictionary", { enumerable: true, get: function () { return core_1.Dictionary; } });
