// @ton/sandbox/dist/metric/gasReportTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gasReportTable = void 0;
const table_1 = require("table");
const deltaResult_1 = require("./deltaResult");
const collectMetric_1 = require("./collectMetric");
const gapCellWidth = 1;
const maxCellWidth = 25;
const border = {
    topBody: '─',
    topJoin: '┬',
    topLeft: '┌',
    topRight: '┐',
    bottomBody: '─',
    bottomJoin: '┴',
    bottomLeft: '└',
    bottomRight: '┘',
    bodyLeft: '│',
    bodyRight: '│',
    bodyJoin: '│',
    joinBody: '─',
    joinLeft: '├',
    joinRight: '┤',
    joinJoin: '┼',
};
function wrap(value, max) {
    if (value.length > max) {
        return `${value.slice(0, max - 3)}...`;
    }
    return value;
}
function prepareResult(list, color) {
    list = list.sort((0, collectMetric_1.sortByCreatedAt)(true));
    const flat = (0, deltaResult_1.toFlatDeltaResult)(list);
    const rows = [];
    const widthCols = [];
    for (const [contract, method, ...values] of flat.rows) {
        const row = [wrap(contract, maxCellWidth), wrap(method, maxCellWidth)];
        widthCols[0] = Math.max(widthCols[0] ?? 0, row[0].length);
        widthCols[1] = Math.max(widthCols[1] ?? 0, row[1].length);
        values.forEach((metric, idx) => {
            const value = color ? color(metric) : metric.value;
            row.push(value);
            const colIdx = idx + 2;
            widthCols[colIdx] = Math.max(widthCols[colIdx] ?? 0, metric.value.length);
        });
        rows.push(row);
    }
    const headers = flat.header;
    for (let i = 0; i < headers.length; i++) {
        widthCols[i] = Math.max(widthCols[i] ?? 0, headers[i].length + gapCellWidth);
    }
    const groupIndex = {};
    let current = 0;
    while (current < rows.length) {
        const contract = rows[current][0];
        let count = 1;
        while (rows[current + count]?.[0] === contract) {
            count++;
        }
        groupIndex[contract] = { index: current, size: count };
        current += count;
    }
    return {
        labels: list.map((s) => s.label),
        headers,
        widthCols,
        rows,
        groupIndex,
    };
}
function gasReportTable(list, color) {
    const result = prepareResult(list, color);
    if (result.rows.length < 1) {
        return 'No data available';
    }
    const columns = [];
    for (let i = 0; i < result.headers.length; i++) {
        columns.push({ alignment: 'center', verticalAlignment: 'middle', width: result.widthCols[i] });
    }
    const spanningCells = [
        { col: 0, row: 0, rowSpan: 2, verticalAlignment: 'middle' },
        { col: 1, row: 0, rowSpan: 2, verticalAlignment: 'middle' }, // Method title
    ];
    for (const group of Object.values(result.groupIndex)) {
        // rowSpan for Contract name
        spanningCells.push({ col: 0, row: group.index + 2, rowSpan: group.size, verticalAlignment: 'middle' });
    }
    const data = [[], ['', '']];
    data[0].push(...result.headers.slice(0, 2));
    data[1].push(...result.headers.slice(2));
    const metricCount = (result.headers.length - 2) / list.length;
    let labelTitleOffset = 2;
    for (const label of result.labels) {
        spanningCells.push({
            col: labelTitleOffset,
            row: 0,
            colSpan: metricCount,
            verticalAlignment: 'middle',
        });
        labelTitleOffset += metricCount;
        data[0].push(...[label, ...Array.from({ length: metricCount - 1 }, () => '')]);
    }
    data.push(...result.rows);
    const config = {
        columns,
        spanningCells,
        border,
    };
    return (0, table_1.table)(data, config);
}
exports.gasReportTable = gasReportTable;
