// @ton/sandbox/dist/debugger/DebugInfoCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCompiledContract = exports.getDebugInfo = void 0;
const marks_1 = require("./marks");
const environment_1 = require("../utils/environment");
const defaultDebugInfoCache = new Map();
function getDebugInfo(acc) {
    if (acc.account === undefined || acc.account === null) {
        return { uninitialized: true, debugInfo: undefined };
    }
    if (acc.account.storage.state.type !== 'active') {
        return { uninitialized: true, debugInfo: undefined };
    }
    const code = acc.account.storage.state.state.code;
    if (code === undefined || code === null) {
        return { uninitialized: true, debugInfo: undefined };
    }
    const debugInfo = defaultDebugInfoCache.get(code.hash().toString('base64'));
    return { uninitialized: false, debugInfo };
}
exports.getDebugInfo = getDebugInfo;
function registerCompiledContract(code, debugInfo, marks) {
    if ((0, environment_1.isBrowser)()) {
        throw new Error('Debug feature is not supported in browser environment');
    }
    const parsedMarks = (0, marks_1.parseMarks)(marks, code);
    const { locations, globals } = debugInfo;
    const sm = {};
    for (let i = 0; i < locations.length; i++) {
        const di = locations[i];
        const common = {
            // eslint-disable-next-line @typescript-eslint/no-require-imports
            path: require('node:path').resolve(di.file),
            line: di.line,
            function: di.func,
            contextId: di.ctx_id,
            requireContextId: di.req_ctx_id,
        };
        if (di.ret) {
            sm[i] = {
                ...common,
                type: 'return',
            };
        }
        else if (di.try_catch_ctx_id !== undefined) {
            sm[i] = {
                ...common,
                type: 'try_begin',
                catchContextId: di.try_catch_ctx_id,
            };
        }
        else if (di.is_try_end) {
            sm[i] = {
                ...common,
                type: 'try_end',
            };
        }
        else if (di.vars !== undefined) {
            sm[i] = {
                ...common,
                type: 'statement',
                variables: di.vars ?? [],
                firstStatement: di.first_stmt,
            };
        }
        else if (di.branch_true_ctx_id !== undefined) {
            sm[i] = {
                ...common,
                type: 'branch',
                trueContextId: di.branch_true_ctx_id,
                falseContextId: di.branch_false_ctx_id,
            };
        }
        else {
            sm[i] = {
                ...common,
                type: 'context',
            };
        }
    }
    defaultDebugInfoCache.set(code.hash().toString('base64'), {
        sourceMap: sm,
        globals,
        marks: parsedMarks,
    });
    return code;
}
exports.registerCompiledContract = registerCompiledContract;
