// @ton/sandbox/dist/debugger/DebugContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDebugContext = void 0;
const environment_1 = require("../utils/environment");
// only this class is allowed to use outside the 'debugger' directory for browser support
class DebugContext {
    constructor() {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const { debugGetMethod, debugTransaction } = require('./debug');
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const { getDebugInfo } = require('./DebugInfoCache');
        this.debugGetMethod = debugGetMethod;
        this.debugTransaction = debugTransaction;
        this.getDebugInfo = getDebugInfo;
    }
}
let debugContext = undefined;
function getDebugContext() {
    if ((0, environment_1.isBrowser)()) {
        throw new Error('Debug feature is not supported in browser environment');
    }
    if (!debugContext) {
        debugContext = new DebugContext();
    }
    return debugContext;
}
exports.getDebugContext = getDebugContext;
